versioned_docs/version-2021sp/lecture5.md

```diff
@@ -3,8 +3,211 @@ id: lecture5
+    <div>You can reach to {name} at
+      <a href={githubLink}>{githubLink}</a>
+    </div>
+  </div>
+);
+
+export default ContactCard;
+```
+
+For this course, we prefer that you use arrow function syntax.
+
+### `index.tsx`
+
+```tsx title="index.tsx"
+import React from 'react';
+import ReactDOM from 'react-dom';
+import ContactCard from './ContactCard';
+
+const link = 'https://github.com/cornell-dti';
+const root = document.getElementById('root');
+
+ReactDOM.render(<ContactCard name="Cornell DTI" githubLink={link} />, root);
+```
+
+In a React app, `index.tsx` is usually the entry point. In the code above, it
+renders your component into the DOM element that has id `root`.
+
+## Stateful Component
+
+Imagine you are writing a contacts app and you need to implement an editor.
+
+Unlike the previous components, you need to maintain state. In React, you will
+need `hooks`.
+
+```tsx
+import { useState, ChangeEvent } from 'react';
+
+const NewContact = () => {
+  // name is the variable for the state, setName is the function you can use
+  // to change the state.
+  const [name, setName] = useState('');
+  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
+    // To extract the value from input box, use the following line.
+    const n = event.currentTarget.value;
+    setName(n);
+  };
+  return (
+    <div>
+      <div>Name: {name}</div>
+      <input
+        type="text"
+        placeholder="enter the name here"
+        value={name}
+        onChange={handleChange}
+      />
+    </div>
+  );
+};
+
+export default NewContact;
+```
+
+## Rendering lists
+
+You may want to render a list of `YourAwesomeComponent`. Here are some examples
+to show how you can achieve this in different ways.
+
+```tsx
+import React from 'react';
+// Suppose you have a ContactCard component defined there.
+import ContactCard from './ContactCard';
+
+type Contact = { name: string; githubLink: string };
+
+const data: Contact[] = [
+  { name: 'Jason', githubLink: 'www.github.com/guessJason' },
+  { name: 'Peter', githubLink: 'www.github.com/peterIsCool' },
+  { name: 'Enoch', githubLink: 'www.github.com/eno' },
+];
+
+const ListBySimpleMap = () => (
+  <div>
+    {data.map((contact: Contact) => (
+      <ContactCard
+        key={contact.name}
+        name={contact.name}
+        githubLink={contact.githubLink}
+      />
+    ))}
+  </div>
+);
+
+const ListBySimpleMapWithObjectDestructing = () => (
+  <div>
+    {data.map(({ name, githubLink }) => (
+      <ContactCard key={name} name={name} githubLink={githubLink} />
+    ))}
+  </div>
+);
+
+const ListBySimpleMapWithSpread = () => (
+  <div>
+    {data.map((contact: Contact) => (
+      <ContactCard key={contact.name} {...contact} />
+    ))}
+  </div>
+);
+```
+
+Note that we always need a `key` prop. Without this, React will give you
+warnings in the console. React needs a unique `key` for each item in the list to
+help it avoid rerendering everything when only one item in the list changes
```

Comment:
Maybe note that you should only use `name` as the key (as done above) if you know that the `name` property is unique. If there are multiple objects with the same name in the list and you used it as the key, it would trip React up.

Reply:
thank you